source/validate_cfg.cpp

```diff
@@ -242,6 +242,25 @@ spv_result_t StructuredControlFlowChecks(
                    _.getIdName(merge->id()), "is not post dominated by");
       }
     }
+
+    // Check that for all non-header blocks, all predecessors are within this
+    // construct.
+    Construct::ConstructBlockSet construct_blocks = construct.blocks(function);
```

Comment:
This constructs the block list for each construct, which will traverse all the blocks in the nested constructs as well.
There's a potential for somewhat pathological behaviour (quadratic?) but dependent heavily on construct nesting depth.
I think this is ok and alternatives would involve state (memoization).
Ok for now.